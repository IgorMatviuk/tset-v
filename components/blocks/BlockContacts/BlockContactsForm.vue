<template>
  <div class="contacts-form">
    <div class="contacts-form__title">
      {{ $t('contacts.titles.form') }}
    </div>
    <form novalidate @submit.prevent="submitContactsForm">
      <div class="contacts-form__row">
        <BFormGroup
          class="contacts-form__group contacts-form__group--name custom-field-group"
          :label="$t('form.labels.first_name')"
          label-for="contactsFormName"
          label-class="contacts-form__label custom-label custom-label--required"
          label-cols="12"
        >
          <BFormInput
            id="contactsFormName"
            v-model.trim="$v.formData.name.$model"
            :class="{ error: $v.formData.name.$error }"
            class="contacts-form__input"
            type="text"
            :trim="true"
            :placeholder="$t('form.placeholders.first_name')"
          />
          <template v-if="$v.formData.name.$error">
            <div
              v-if="getShowModelError('name', 'required')"
              class="error-message"
            >
              {{ $t(`form.errors.first_name.required`) }}
            </div>
            <div
              v-else-if="getShowModelError('name', 'minLength')"
              class="error-message"
            >
              {{ $t(`form.errors.first_name.minLength`) }}
            </div>
            <div
              v-else-if="getShowModelError('name', 'onlyLetters')"
              class="error-message"
            >
              {{ $t(`form.errors.first_name.onlyLetters`) }}
            </div>
          </template>
        </BFormGroup>
        <BFormGroup
          class="contacts-form__group custom-field-group"
          :label="$t('form.labels.phone')"
          label-for="contactsFormPhone"
          label-class="contacts-form__label custom-label custom-label--required"
          label-cols="12"
        >
          <client-only>
            <vue-tel-input
              ref="tel"
              v-model.trim="$v.formData.phone.$model"
              v-bind="phoneOptions"
              :wrapper-classes="
                $v.formData.phone.$error
                  ? 'form-control-phone form-control error'
                  : 'form-control-phone form-control'
              "
            />
          </client-only>
          <template v-if="$v.formData.phone.$error">
            <div
              v-if="getShowModelError('phone', 'required')"
              class="error-message"
            >
              {{ $t(`form.errors.phone.required`) }}
            </div>
            <div
              v-else-if="getShowModelError('phone', 'phone')"
              class="error-message"
            >
              {{ $t(`form.errors.phone.phone`) }}
            </div>
          </template>
        </BFormGroup>
        <BFormGroup
          class="contacts-form__group contacts-form__group--email custom-field-group"
          :label="$t('form.labels.email')"
          label-for="contactsFormEmail"
          label-class="contacts-form__label custom-label custom-label--required"
          label-cols="12"
        >
          <BFormInput
            id="contactsFormEmail"
            v-model.trim="$v.formData.email.$model"
            :class="{ error: $v.formData.email.$error }"
            class="contacts-form__input"
            type="email"
            :trim="true"
            :placeholder="$t('form.placeholders.email')"
          />
          <template v-if="$v.formData.email.$error">
            <div
              v-if="getShowModelError('email', 'required')"
              class="error-message"
            >
              {{ $t(`form.errors.email.required`) }}
            </div>
            <div
              v-else-if="getShowModelError('email', 'email')"
              class="error-message"
            >
              {{ $t(`form.errors.email.email`) }}
            </div>
          </template>
        </BFormGroup>
        <BFormGroup
          class="contacts-form__group custom-field-group"
          :label="$t('form.labels.comment')"
          label-for="contactsFormComment"
          label-class="contacts-form__label custom-label"
          label-cols="12"
        >
          <BFormTextarea
            id="contactsFormComment"
            v-model.trim="$v.formData.comment.$model"
            :class="{ error: $v.formData.comment.$error }"
            class="contacts-form__input"
            :trim="true"
            :placeholder="$t('form.placeholders.comment')"
          />
          <template v-if="$v.formData.comment.$error">
            <div
              v-if="getShowModelError('comment', 'onlyText')"
              class="error-message"
            >
              {{ $t(`form.errors.comment.onlyText`) }}
            </div>
          </template>
        </BFormGroup>
      </div>
      <ButtonLoading
        :text="submitText"
        :is-loading="isLoading"
        type="submit"
        class="contacts-form__submit"
      />
    </form>
  </div>
</template>

<script>
import { validationMixin } from 'vuelidate'
import { required, email, minLength } from 'vuelidate/lib/validators'
import { mutationTypes as modalMutationTypes } from '@/store/modal'

import ButtonLoading from '@/components/shared/ButtonLoading'

const onlyLetters = (value) =>
  /^([A-zА-яЁёА-Яа-яЁёЇїІіЄєҐґ\-'])+$/gm.test(value)
const onlyText = (value) =>
  /^([A-zА-яЁёА-Яа-яЁёЇїІіЄєҐґ0-9\-\\.,;:?!'"()\s]||)+$/gm.test(value)

export default {
  components: {
    ButtonLoading
  },
  mixins: [validationMixin],
  data() {
    return {
      phoneOptions: {
        inputId: 'contactsFormPhone',
        inputClasses: 'contacts-form__input form-control',
        placeholder: 'Ваш телефон',
        validCharactersOnly: true,
        mode: 'international'
      },
      isLoading: false,
      formData: {
        name: '',
        phone: '',
        email: '',
        comment: ''
      }
    }
  },
  computed: {
    submitText() {
      return this.$t('buttons.submit')
    }
  },
  methods: {
    submitContactsForm() {
      this.$v.$touch()

      /* istanbul ignore else */
      if (!this.$v.$invalid) {
        this.isLoading = true

        // TODO: submit form
        setTimeout(() => {
          this.isLoading = false

          this.$store.commit(`modal/${modalMutationTypes.SET_MODAL_META}`, {
            type: 'message',
            title: this.$t('modals.contactsMessage.title'),
            content: this.$t('modals.contactsMessage.content')
          })

          this.$store.commit(
            `modal/${modalMutationTypes.SET_MODAL_VISIBILITY_TOGGLE}`,
            { type: 'message', status: true }
          )
        }, 1000)
      }
    },
    getShowModelError(fieldname, validatorName) {
      const validator = this.$v.formData[fieldname][validatorName]
      return typeof validator !== 'undefined' && !validator
    }
  },
  validations: {
    formData: {
      name: {
        required,
        minLength: minLength(4),
        onlyLetters
      },
      phone: {
        required,
        phone() {
          const tel = this.$refs.tel
          return tel.phoneObject.isValid
        }
      },
      email: {
        email,
        required
      },
      comment: {
        onlyText
      }
    }
  }
}
</script>

<style lang="scss">
.contacts-form {
  &__title {
    margin-bottom: 17px;
    font-size: 14px;
    line-height: 18px;
    text-transform: uppercase;
    letter-spacing: 0.05em;
  }

  &__row {
    margin-bottom: 30px;
  }

  &__group {
    margin-bottom: 10px;
  }

  &__submit {
    display: block;
    min-width: 250px;
    margin-right: auto;
    margin-left: auto;
  }

  @include media-breakpoint-up(md) {
    &__title {
      margin-bottom: 12px;
    }

    &__row {
      @include make-row();

      align-content: flex-end;
      align-items: flex-end;
      margin-bottom: 37px;
    }

    &__group {
      @include make-col(6);
      @include make-col-ready();

      margin-bottom: 8px;

      &--name {
        order: -2;
      }

      &--email {
        order: -1;
      }
    }

    &__submit {
      margin: 0;
    }
  }

  @include media-breakpoint-up(lg) {
    &__group {
      flex: 0 0 260px;
      max-width: 260px;
    }
  }

  @include media-breakpoint-up(xl) {
    &__row {
      margin-bottom: 34px;
    }

    &__group {
      flex: 0 0 310px;
      max-width: 310px;
    }
  }

  @include media-breakpoint-up(xxl) {
    &__row {
      margin-bottom: 19px;
    }

    &__group {
      @include make-col(6);
    }
  }
}
</style>
