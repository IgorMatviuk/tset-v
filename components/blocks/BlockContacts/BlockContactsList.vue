<template>
  <ul class="contacts-list">
    <li class="contacts-list__item">
      <svg-icon
        name="phone"
        class="contacts-list__icon"
        width="16"
        height="16"
      />
      <span class="contacts-list__description">
        <span class="contacts-list__title">
          {{ $t('contacts.titles.phone') }}
        </span>
        <a
          :href="`tel:${contacts.phone}` | convertPhone"
          class="contacts-list__link"
        >
          {{ contacts.phone }}
        </a>
      </span>
    </li>
    <li class="contacts-list__item">
      <svg-icon
        name="mail"
        class="contacts-list__icon"
        width="16"
        height="16"
      />
      <span class="contacts-list__description">
        <span class="contacts-list__title">
          {{ $t('contacts.titles.email') }}
        </span>
        <a :href="`mailto:${contacts.email}`" class="contacts-list__link">
          {{ contacts.email }}
        </a>
      </span>
    </li>
    <li class="contacts-list__item">
      <svg-icon
        name="headset"
        class="contacts-list__icon"
        width="16"
        height="16"
      />
      <span class="contacts-list__description">
        <span class="contacts-list__title">
          {{ $t('contacts.titles.schedule') }}
        </span>
        <span class="contacts-list__schedule">{{ contacts.schedule }}</span>
      </span>
    </li>
  </ul>
</template>

<script>
import phoneConversion from '@/mixins/phone-conversion'

export default {
  mixins: [phoneConversion],
  computed: {
    contacts() {
      return this.$store.state.contacts
    }
  }
}
</script>

<style lang="scss">
.contacts-list {
  @include list-unstyled();

  &__item {
    display: flex;

    &:not(:last-of-type) {
      margin-bottom: 19px;
    }
  }

  &__icon {
    flex: 0 0 16px;
    width: 16px;
    height: 16px;
    margin-top: 3px;
    color: $primary;
  }

  &__description {
    display: flex;
    flex-direction: column;
    padding-left: 18px;
  }

  &__title {
    margin-bottom: 4px;
    font-size: 11px;
    line-height: 14px;
    color: $gray-dark-1;
  }

  &__link,
  &__schedule {
    font-size: 13px;
    font-weight: 500;
    line-height: 16px;
    letter-spacing: 0.05em;
  }

  @include media-breakpoint-up(md) {
    display: flex;
    flex-wrap: wrap;

    &__item {
      @include make-col(4);
      @include make-col-ready();

      flex-direction: column;
    }

    &__description {
      padding-left: 0;
    }

    &__icon {
      margin-top: 0;
      margin-bottom: 14px;
    }

    &__title {
      margin-bottom: 7px;
    }
  }
}
</style>
