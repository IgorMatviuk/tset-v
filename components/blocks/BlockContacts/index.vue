<template>
  <div class="contacts">
    <div class="contacts__background-wrapper">
      <div class="contacts__color-background"></div>
      <picture>
        <source
          data-srcset="/img/contacts-page/contacts_background__xxl.webp"
          srcset="/img/stub.png"
          type="image/webp"
          media="(min-width: 1500px)"
        />
        <source
          data-srcset="/img/contacts-page/contacts_background__xxl.jpg"
          srcset="/img/stub.png"
          type="image/jpeg"
          media="(min-width: 1500px)"
        />
        <source
          data-srcset="/img/contacts-page/contacts_background__xl.webp"
          srcset="/img/stub.png"
          type="image/webp"
          media="(min-width: 1200px)"
        />
        <source
          data-srcset="/img/contacts-page/contacts_background__xl.jpg"
          srcset="/img/stub.png"
          type="image/jpeg"
          media="(min-width: 1200px)"
        />
        <source
          data-srcset="/img/contacts-page/contacts_background__md.webp"
          srcset="/img/stub.png"
          type="image/webp"
          media="(min-width: 768px)"
        />
        <source
          data-srcset="/img/contacts-page/contacts_background__md.jpg"
          srcset="/img/stub.png"
          type="image/jpeg"
          media="(min-width: 768px)"
        />
        <img
          data-src=""
          data-object-fit="cover"
          data-object-position="top"
          src="/img/stub.png"
          alt="contacts page bockground"
          class="contacts__background-image lazyload"
          width="100"
          height="100"
          loading="lazy"
        />
      </picture>
    </div>
    <BContainer class="contacts__container">
      <div class="contacts__info-background"></div>
      <h1 class="contacts__title">{{ $t('contacts.titles.main') }}</h1>
      <BlockContactsList class="contacts__list" />
      <BlockContactsForm class="contacts__form" />
    </BContainer>
  </div>
</template>

<script>
import BlockContactsList from './BlockContactsList'
import BlockContactsForm from './BlockContactsForm'

export default {
  components: {
    BlockContactsList,
    BlockContactsForm
  }
}
</script>

<style lang="scss">
.contacts {
  padding-top: 56px;
  padding-bottom: 61px;
  overflow: hidden;

  &__background-wrapper {
    display: none;
  }

  &__title {
    margin-bottom: 25px;
    font-size: 16px;
    font-weight: 400;
    line-height: 20px;
    text-transform: uppercase;
    letter-spacing: 0.05em;
  }

  &__list {
    margin-bottom: 44px;
  }

  @include media-breakpoint-up(md) {
    position: relative;
    display: flex;
    min-height: calc(100vh - 550px);
    padding-top: 103px;
    padding-bottom: 0;

    &__container {
      position: relative;
      padding-top: 26px;
      padding-bottom: 34px;
    }

    &__background-wrapper {
      position: absolute;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      display: block;
    }

    &__background-image {
      width: 100%;
      max-width: 560px;
      height: 100%;
      object-fit: cover;
      object-position: top;
    }

    &__color-background {
      position: absolute;
      top: 0;
      right: 0;
      bottom: 0;
      left: 42.5%;
      background-color: $gray-light-3;
    }

    &__info-background {
      position: absolute;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      margin-right: -50vw;
      margin-left: 27.05%;
      background-color: $white;
    }

    &__title,
    &__form {
      position: relative;
      margin-left: 29.1%;
    }

    &__list {
      position: relative;
      margin-bottom: 25px;
      margin-left: 27%;
    }

    &__title {
      margin-bottom: 25px;
    }
  }

  @include media-breakpoint-up(lg) {
    min-height: calc(100vh - 440px);
    padding-top: 97px;

    &__info-background {
      margin-left: 34.9%;
    }

    &__title,
    &__form {
      margin-left: 37.7%;
    }

    &__list {
      margin-left: 36.1%;
    }
  }

  @include media-breakpoint-up(xl) {
    min-height: calc(100vh - 390px);
    padding-top: 140px;

    &__color-background {
      left: 49%;
    }

    &__background-image {
      max-width: 49%;
    }

    &__container {
      padding-top: 30px;
      padding-bottom: 38px;
    }

    &__info-background {
      margin-left: 42.9%;
    }

    &__title,
    &__form {
      margin-left: 45.7%;
    }

    &__title {
      margin-bottom: 39px;
    }

    &__list {
      margin-bottom: 36px;
      margin-left: 44.45%;
    }
  }

  @include media-breakpoint-up(xxl) {
    padding-top: 173px;

    &__color-background {
      left: 49.22%;
    }

    &__background-image {
      max-width: 49.22%;
    }

    &__container {
      padding-top: 40px;
      padding-bottom: 78px;
    }

    &__info-background {
      margin-left: 48.4%;
    }

    &__title {
      margin-bottom: 27px;
    }

    &__title,
    &__form {
      margin-left: 51.2%;
    }

    &__list {
      margin-bottom: 46px;
      margin-left: 50.2%;
    }
  }
}
</style>
