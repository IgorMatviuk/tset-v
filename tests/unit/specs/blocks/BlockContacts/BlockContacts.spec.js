import { shallowMount, createLocalVue } from '@vue/test-utils'
import { BContainer } from 'bootstrap-vue'

import BlockContacts from '@/components/blocks/BlockContacts'

const localVue = createLocalVue()

localVue.component('BContainer', BContainer)

describe('BlockContacts', () => {
  test('renders contacts block', () => {
    const wrapper = shallowMount(BlockContacts, {
      localVue,
      mocks: {
        $t: () => ''
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })
})
