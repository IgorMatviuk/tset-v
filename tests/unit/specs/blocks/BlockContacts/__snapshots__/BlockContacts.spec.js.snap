// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockContacts renders contacts block 1`] = `
<div
  class="contacts"
>
  <div
    class="contacts__background-wrapper"
  >
    <div
      class="contacts__color-background"
    />
     
    <picture>
      <source
        data-srcset="/img/contacts-page/contacts_background__xxl.webp"
        media="(min-width: 1500px)"
        srcset="/img/stub.png"
        type="image/webp"
      />
       
      <source
        data-srcset="/img/contacts-page/contacts_background__xxl.jpg"
        media="(min-width: 1500px)"
        srcset="/img/stub.png"
        type="image/jpeg"
      />
       
      <source
        data-srcset="/img/contacts-page/contacts_background__xl.webp"
        media="(min-width: 1200px)"
        srcset="/img/stub.png"
        type="image/webp"
      />
       
      <source
        data-srcset="/img/contacts-page/contacts_background__xl.jpg"
        media="(min-width: 1200px)"
        srcset="/img/stub.png"
        type="image/jpeg"
      />
       
      <source
        data-srcset="/img/contacts-page/contacts_background__md.webp"
        media="(min-width: 768px)"
        srcset="/img/stub.png"
        type="image/webp"
      />
       
      <source
        data-srcset="/img/contacts-page/contacts_background__md.jpg"
        media="(min-width: 768px)"
        srcset="/img/stub.png"
        type="image/jpeg"
      />
       
      <img
        alt="contacts page bockground"
        class="contacts__background-image lazyload"
        data-object-fit="cover"
        data-object-position="top"
        data-src=""
        height="100"
        loading="lazy"
        src="/img/stub.png"
        width="100"
      />
    </picture>
  </div>
   
  <bcontainer-stub
    class="contacts__container"
    tag="div"
  >
    <div
      class="contacts__info-background"
    />
     
    <h1
      class="contacts__title"
    >
      
    </h1>
     
    <blockcontactslist-stub
      class="contacts__list"
    />
     
    <blockcontactsform-stub
      class="contacts__form"
    />
  </bcontainer-stub>
</div>
`;
